### ISSUE TITLE
Building OpenSSL statically: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking

### ISSUE THREAD

**Issue by victorpaleologue** [messense/homebrew-macos-cross-toolchains#12]
Hello,
I am trying to build OpenSSL in static using the i686 toolchain by calling something like, in short:
```./Configure linux-x86 -static```
It seems it is well using `libc.a`, but the following warning is produced:
```Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking```
As a result, my binary depends on glibc at runtime, and since the version of glibc on my target platform differs from the one of the toolchain, the executable I produced is not compatible.

According to [this issue](https://github.com/openssl/openssl/issues/14917), which exhibits a similar issue, it is a problem with how glibc is compiled. For instance it is fixed [here](https://github.com/0neday/buildroot/commit/7409896601fb2c847bdccf0db9fcb6aa75702789) by adding the flag `--enable-static-nss` when building statically.
Perhaps that could do the job here?

In the meantime I will try to bring glibc along.

**Comment by messense** [comment]
Why not just use musl libc if you want fully statically linked binary?

**Comment by victorpaleologue** [comment]
I've just tried and everything builds. But the resulting binary cannot be executed, it fails with a `no such file or directory` message:
```sh
$ ./my_bin
-bash: ./my_bin: No such file or directory
```
Using `ldd` on it also produces this error.
Looking at the executable using `readelf` shows something consistent.
Is that how `musl` tells me we're missing a dynamic library?

**Comment by messense** [comment]
Can you write down the build steps you used? And how does `file my_bin` outputs?

**Comment by messense** [comment]
I think that cmake configuration dynamic links musl, you should try this: https://stackoverflow.com/a/61815350

**Comment by victorpaleologue** [comment]
I am already using `-static-libgcc` `-static-libstdc++`, but I cannot use `-static` because the third party library I use have to be compiled into a dynamic library. Is there a way to bring `ld-musl-i386.so.1` to the platform? Or can I tell the toolchain to use another `ld-linux.so` coming from the platform?

**Comment by messense** [comment]
I don't know if it's possible to bring `ld-musl-i386.so.1` to the platform, I've never tried. 

> the third party library I use have to be compiled into a dynamic library

In that case I think maybe it's better to  just upgrade the glibc version of your target platform.

**Comment by messense** [comment]
And if you really need to target a lower glibc version, [`zig cc` ](https://andrewkelley.me/post/zig-cc-powerful-drop-in-replacement-gcc-clang.html) is worth a shot.

**Comment by victorpaleologue** [comment]
Ok, that's it, I managed to produce a static version of the third-party library.
And using the `musl` toolchain, OpenSSL does not require us to keep `glibc.so` around.
Altogether, I managed to make an executable that works on the target platform, that is stuck with an old `glibc`!
Thank you so much.

My conclusion is that this issue is rather about OpenSSL and other libraries that depend on `dl`, and that for those, using the `musl` toolchains is more portable. There is not much to do about it in this repository.
